### PR TITLE
Feat/multiclass training

### LINKED ISSUE #48: Migrate Multiclass Semantic Segmentation
## Description
Currently, the URC repository does not utilize machine learning of any kind for obstacle detection. A good way to kickstart our ML efforts would be to migrate the vision based perception strategy from the IGVC repository.

The vision based perception strategy in the IGVC repository relies on a [UNET](https://arxiv.org/abs/1505.04597) with an [EfficientNet](https://arxiv.org/abs/1905.11946) encoder to perform semantic segmentation on input images. The neural net for the IGVC competition was trained to classify pixels as line, barrel, or ground. For the URC, the network will probably have to be retrained to classify obstacles for the URC competition. Retraining will involve some effort to label images.

Our encoder is pretrained on the [ImageNet](https://en.m.wikipedia.org/wiki/ImageNet) dataset which allows us to utilize transfer learning to train our model. Networks trained from scratch need a massive amounts of data and compute. Even with the appropriate dataset and compute you would need to wait days or weeks to train such a model. Using this strategy we can get away with using 500ish images (versus millions of images needed if training from scratch) to fine tune the weights of the pretrained encoder and achieve highly accurate results.

## Requirements

 - Migrate [multiclass_segmentation.py](https://github.com/RoboJackets/igvc-software/blob/c4de30fc741c9fb3d0923596bb2263d947eec3dd/igvc_perception/src/multiclass_segmentation/multiclass_segmentation.py) and the corresponding [multiclass segmentation folder](https://github.com/RoboJackets/igvc-software/tree/master/igvc_perception/src/multiclass_segmentation) 
 - Migrate [multiclass_segmentation.launch](https://github.com/RoboJackets/igvc-software/blob/c4de30fc741c9fb3d0923596bb2263d947eec3dd/igvc_perception/launch/multiclass_segmentation.launch)
 - Successfully launches

### PR DESCRIPTION
**Pull request by a-stickan**
# Description

Sets up a basic framework for training the multiclass segmentation model. This is not fully functional yet, will be completed when the rover is fully constructed and sample images can be taken.

This PR does the following:
- Copies over python training files from IGVC
- Adds documentation for `multiclass_segmentation`
- Adds training instructions for
  - Google Colab
  - Jupyter notebook
  - Local runs

Fixes #48 

# Testing steps
1. Read over the documentation
2. Read over the training instructions

Expectation: No typos/formatting errors

# Self Checklist
- [x] I have formatted my code using `ament_uncrustify --reformat`
- [x] I have tested that the new behavior works

### CODE REVIEW

File: urc_perception/multiclass_segmentation/train.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    args = vars(get_args())
+
+    train_images_path = args["train_images"]
+    train_masks_path = args["train_masks"]
+    test_images_path = args["test_images"]
+    test_masks_path = args["test_masks"]
+
+    # print out yaml file configuration
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    yaml_path = os.path.join(dir_path, "config/igvc.yaml")
```
Comment (VAM7686) [inline urc_perception/multiclass_segmentation/train.py:38]: This should be `urc.yaml`

**Review by VAM7686 — CHANGES_REQUESTED** [review]
Looks good to me. Just a small change needed.

File: urc_perception/multiclass_segmentation/train.py (hunk 1 above)
Comment (a-stickan) [reply urc_perception/multiclass_segmentation/train.py]: Just pushed a change removing the last few IGVC references, few must have slipped by me when I was looking the code over

File: urc_perception/multiclass_segmentation/README.md [hunk 2]
```diff
@@ -80,7 +80,7 @@ You can alternatively use a google colab to train your model. Here is a brief in
 
 2.  Open the [ConvertImages_SplitData.ipynb](https://colab.research.google.com/drive/15T2XQjBKMh92A3XlsfR-kRX-PFVN9rl4?usp=sharing) Google Colab Notebook and follow the steps to convert your dataset into the appropriate format and create train/test splits.
     
-3.  Open the [urc_multiunet.ipynb](https://colab.research.google.com/drive/1bb9TRCNWBgV8-EiqzhjlSQM_yny9DXzt?usp=sharing) Google Colab Notebook and follow the steps to train the network. Make sure to adjust the config file in igvc_perception/src/multiclass_segmentation/config/igvc.yaml for desired training parameters and to specify model save location.
+3.  Open the [urc_multiunet.ipynb](https://colab.research.google.com/drive/1bb9TRCNWBgV8-EiqzhjlSQM_yny9DXzt?usp=sharing) Google Colab Notebook and follow the steps to train the network. Make sure to adjust the config file in urc_perception/multiclass_segmentation/config/urc.yaml for desired training parameters and to specify model save location.
```
Comment (VAM7686) [inline urc_perception/multiclass_segmentation/README.md:83]: Why is multiclass_segmentation not in a src folder? Shouldnt it follow the same format as all other nodes?

**Review by VAM7686 — CHANGES_REQUESTED** [review]
just a small question on the folder structure?

File: urc_perception/multiclass_segmentation/README.md (hunk 2 above)
Comment (a-stickan) [reply urc_perception/multiclass_segmentation/README.md:83]: I personally believe that multiclass segmentation should be it's own directory because it is more than just the one node. It has its own README, lots of training files, seperate config files, etc. In my opinion anything that complex really deserves its own directory, I view `src` as a holder for more prototypical nodes.

Comment (VAM7686) [reply urc_perception/multiclass_segmentation/README.md:83]: Ok sounds good. Just wanted to clarify this point.

**Review by VAM7686 — APPROVED** [review]
looks good to me!
